File: Source/2017/Day12_2.go
package main

import (
    "bufio"
    "fmt"
    "log"
    "os"
    "strconv"
    "strings"
)

type Node struct {
    Id int
    Neighbours map[int]*Node
}

func main() {
    file, ok := os.Open("Day12.txt")
    if ok != nil {
        log.Fatal("Could not open input file")
    }
    defer file.Close()

    nodes := map[int]*Node {}

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        split := strings.Split(scanner.Text(), " <-> ")
        nodeId, err := strconv.Atoi(split[0])
        if err != nil {
            log.Fatal(err)
        }

        var node *Node
        if existingNode, ok := nodes[nodeId]; ok {
            node = existingNode
        } else {
            node = &Node { Id: nodeId, Neighbours: map[int]*Node {} }
            nodes[nodeId] = node
        }

        for _, neighbour := range strings.Split(split[1], ", ") {
            neighId, err := strconv.Atoi(neighbour)
			if err != nil {
				log.Fatal(err)
            }

            var neighbour *Node
            if existing, ok := nodes[neighId]; ok {
                neighbour = existing
            } else {
                neighbour = &Node { Id: neighId, Neighbours: map[int]*Node {} }
                nodes[neighId] = neighbour
            }

            if _, ok := neighbour.Neighbours[nodeId]; ok { continue }
            if _, ok := node.Neighbours[neighId]; ok { continue }

            neighbour.Neighbours[nodeId] = node
            node.Neighbours[neighId] = neighbour
        }
    }

    members := make([]map[int]*Node, 0)
    queue := make([]*Node, 0)

    for _, node := range nodes {
        nodeHasGroup := false
        for _, group := range members {
            if _, ok := group[node.Id]; ok {
                nodeHasGroup = true
                break
            }
        }
        if nodeHasGroup { continue }

        group := make(map[int]*Node, 0)
        members = append(members, group)
        queue = append(queue, node)

        for len(queue) > 0 {
            dequeued := queue[0]
            queue = queue[1:]

            for _, neigh := range dequeued.Neighbours {
                if _, ok := group[neigh.Id]; ok { continue }
                group[neigh.Id] = neigh
                queue = append(queue, neigh)
            }
        }
    }

    fmt.Println(len(members))

    // Push
    // queue = append(queue, 1)

    // Top (just get next element, don't remove it)
    // x := queue[0]

    // Discard top element
    // queue = queue[1:]

    // Is empty ?
    //if len(queue) == 0 { ... }

    //PrintAllNodes(nodes)
}

func PrintAllNodes(nodes map[int]*Node) {
    for _, node := range nodes {
        fmt.Printf("[%d] ", node.Id)
        for _, neigh := range node.Neighbours {
            fmt.Printf ("%d ", neigh.Id)
        }
        fmt.Println()
    }
}

File: archive/Source/2020/Day12_2.py
file = open("Day12.txt")

x = 0
y = 0
wx = 10
wy = -1

directions = ['E', 'S', 'W', 'N']

vectors = {
    'E': (1, 0),
    'W': (-1, 0),
    'N': (0, -1),
    'S': (0, 1)
}

for line in file:
    op = line[0]
    val = int(line[1:-1])

    if (op == 'F'):
        x += wx*val
        y += wy*val

    elif (op in directions):
        (xmod, ymod) = vectors[op]
        wx += val*xmod
        wy += val*ymod

    if (op == 'L' or op == 'R'):
        val = val // 90
        if (op == 'L'): val = -val
        val = (4 + val) % 4
        for i in range(val):
            t = wx
            wx = -wy
            wy = t

    print(f"{line[:-1]}: ({x}, {y}) [{wx}, {wy}]")

print (f"({x}, {y}) => {abs(x)+abs(y)}")

